File: frontend/src/services/itemDiagnosticoService.ts
import api from './api'
import type {
  IItemDiagnostico,
  IItemDiagnosticoFormPayload,
  IItemDiagnosticoListResponse,
} from '@/types/itemDiagnostico'

export interface ItemDiagnosticoListParams {
  page?: number
  per_page?: number
  search?: string
}

export async function getItemsDiagnostico(
  params: ItemDiagnosticoListParams = {},
): Promise<IItemDiagnosticoListResponse> {
  const { data } = await api.get<IItemDiagnosticoListResponse>('/items-diagnostico', { params })
  return data
}

export async function createItemDiagnostico(
  payload: IItemDiagnosticoFormPayload,
): Promise<IItemDiagnostico> {
  const { data } = await api.post<{ data: IItemDiagnostico }>('/items-diagnostico', payload)
  return data.data
}

export async function updateItemDiagnostico(
  id: number,
  payload: IItemDiagnosticoFormPayload,
): Promise<IItemDiagnostico> {
  const { data } = await api.put<{ data: IItemDiagnostico }>(`/items-diagnostico/${id}`, payload)
  return data.data
}

export async function deleteItemDiagnostico(id: number): Promise<void> {
  await api.delete(`/items-diagnostico/${id}`)
}

File: ferias_app/lib/services/item_diagnostico_service.dart
import '../models/item_diagnostico.dart';
import '../models/paginated_response.dart';
import 'api_service.dart';

class ItemDiagnosticoService {
  ItemDiagnosticoService({ApiService? apiService})
    : _apiService = apiService ?? ApiService();

  final ApiService _apiService;

  Future<PaginatedResponse<ItemDiagnostico>> getItemsDiagnostico({
    String? search,
    int page = 1,
    int perPage = 15,
  }) async {
    final response = await _apiService.get<Map<String, dynamic>>(
      '/items-diagnostico',
      queryParameters: <String, dynamic>{
        'search': search?.trim().isEmpty == true ? null : search?.trim(),
        'page': page,
        'per_page': perPage,
      }..removeWhere((key, value) => value == null),
    );

    return PaginatedResponse<ItemDiagnostico>.fromJson(
      response.data ?? <String, dynamic>{},
      ItemDiagnostico.fromJson,
    );
  }

  Future<ItemDiagnostico> createItemDiagnostico({
    required String nombre,
  }) async {
    final response = await _apiService.post<Map<String, dynamic>>(
      '/items-diagnostico',
      data: <String, dynamic>{'nombre': nombre.trim()},
    );

    return _fromResource(response.data);
  }

  Future<ItemDiagnostico> updateItemDiagnostico({
    required int id,
    required String nombre,
  }) async {
    final response = await _apiService.put<Map<String, dynamic>>(
      '/items-diagnostico/$id',
      data: <String, dynamic>{'nombre': nombre.trim()},
    );

    return _fromResource(response.data);
  }

  Future<void> deleteItemDiagnostico(int id) async {
    await _apiService.delete<Map<String, dynamic>>('/items-diagnostico/$id');
  }

  ItemDiagnostico _fromResource(Map<String, dynamic>? payload) {
    final data = payload?['data'];

    if (data is Map) {
      return ItemDiagnostico.fromJson(Map<String, dynamic>.from(data));
    }

    return ItemDiagnostico.fromJson(payload ?? <String, dynamic>{});
  }
}
